# Parkview Regional Medical Center

https://www.parkview.com/patients-and-visitors/billing-and-insurance/request-an-estimate 

Parkview Regional Medical Center & Parkview Hospital Randallia:

https://www.parkview.com/-/media/parkview-media/file/0720wp---2020-new/350868085_parkview-regional-medical-center_standardcharges-(1).ashx 

Following Lehigh Valley.

In [1]:
# Load duckdb, which lets us efficiently load large files
import duckdb

# Load pandas, which lets us manipulate dataframes
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

# Set configrations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.
%config SqlMagic.autopandas = True

%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

# Allow named parameters (python variables) in SQL cells
%config SqlMagic.named_parameters=True

# Connect jupysql to DuckDB using a SQLAlchemy-style connection string. Either connect to an in memory DuckDB, or a file backed db.
%sql duckdb:///:memory:

In [7]:
!wget "https://www.parkview.com/-/media/parkview-media/file/0720wp---2020-new/350868085_parkview-regional-medical-center_standardcharges-(1).ashx" -P /tmp

--2023-09-07 15:27:18--  https://www.parkview.com/-/media/parkview-media/file/0720wp---2020-new/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
Resolving www.parkview.com (www.parkview.com)... 162.209.114.56
Connecting to www.parkview.com (www.parkview.com)|162.209.114.56|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4814647 (4.6M) [application/octet-stream]
Saving to: ‘/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx’

350868085_parkview- 100%[===================>]   4.59M  7.39MB/s    in 0.6s    

2023-09-07 15:27:19 (7.39 MB/s) - ‘/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx’ saved [4814647/4814647]



In [12]:
%%sql
SELECT * FROM read_json_auto('/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx', records=true, maximum_object_size=500000000)

File Summary  \
0  [{'Hospital Name': 'Parkview Regional Medical ...   

                                       Gross Charges  \
0  [{'CDM/EAP': '01240002', 'Description': 'HC GE...   

                      Discounted Cash Pricing Policy  \
0  [{'Policy Line': 'A 30% Medically Uninsured Di...   

   Inpatient De-identified Minimum Negotiated Charge  \
0  [{'MS-DRG': '003', 'Description': 'Ecmo Or Tra...   

   Inpatient De-identified Maximum Negotiated Charge  \
0  [{'MS-DRG': '003', 'Description': 'Ecmo Or Tra...   

                   Inpatient Payer Specific Charge 1  \
0  [{'Payer': 'AETNA', 'MS-DRG': '003', 'Descript...   

                   Inpatient Payer Specific Charge 2  \
0  [{'Payer': 'ANTHEM HMO/PPO/TRADITIONAL', 'MS-D...   

                   Inpatient Payer Specific Charge 3  \
0  [{'Payer': 'CHRISTIAN HEALTHCARE MINISTRIES', ...   

                   Inpatient Payer Specific Charge 4  \
0  [{'Payer': 'CIGNA', 'MS-DRG': '003', 'Descript...   

                   Inpatient Payer Specific Charge 5  ...  \
0  [{'Payer': 'ENCORE', 'MS-DRG': '003', 'Descrip...  ...   

                 Outpatient Payer Specific Charge 64  \
0  [{'Payer': 'KOSCIUSKO COUNTY JAIL', 'APC': '51...   

                 Outpatient Payer Specific Charge 65  \
0  [{'Payer': 'ENCORE-ENCIRCLE', 'APC': '5362', '...   

                 Outpatient Payer Specific Charge 66  \
0  [{'Payer': 'K AND K INSURANCE', 'APC': 5522, '...   

                 Outpatient Payer Specific Charge 67  \
0  [{'Payer': 'MEDICARE MANAGED CARE-BUCKEYE', 'A...   

                 Outpatient Payer Specific Charge 68  \
0  [{'Payer': 'BEACON HEALTH OPTIONS', 'APC': 582...   

                 Outpatient Payer Specific Charge 69  \
0  [{'Payer': 'LAGRANGE COUNTY JAIL', 'APC': 'N80...   

                 Outpatient Payer Specific Charge 70  \
0  [{'Payer': 'MEDICAID MANAGED CARE-ANTHEM', 'AP...   

                 Outpatient Payer Specific Charge 71  \
0  [{'Payer': 'PHP PLATINUM', 'APC': 'N800', 'Des...   

                 Outpatient Payer Specific Charge 72  \
0  [{'Payer': 'PLAIN CHURCH-IMMERGRUN', 'APC': 'N...   

                 Outpatient Payer Specific Charge 73  
0  [{'Payer': 'APOSTROPHE HEALTH', 'APC': 'N801',...  

[1 rows x 140 columns]

## Claude prompt
````
please write a jq command to retrieve the 140 column names from the json file at /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx where the column names are at the top level and followed by lists.
````

In [14]:
!jq -r 'keys | .[]' "/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx"

Discounted Cash Pricing Policy
File Summary
Gross Charges
Inpatient De-identified Maximum Negotiated Charge
Inpatient De-identified Minimum Negotiated Charge
Inpatient Payer Specific Charge 1
Inpatient Payer Specific Charge 10
Inpatient Payer Specific Charge 11
Inpatient Payer Specific Charge 12
Inpatient Payer Specific Charge 13
Inpatient Payer Specific Charge 14
Inpatient Payer Specific Charge 15
Inpatient Payer Specific Charge 16
Inpatient Payer Specific Charge 17
Inpatient Payer Specific Charge 18
Inpatient Payer Specific Charge 19
Inpatient Payer Specific Charge 2
Inpatient Payer Specific Charge 20
Inpatient Payer Specific Charge 21
Inpatient Payer Specific Charge 22
Inpatient Payer Specific Charge 23
Inpatient Payer Specific Charge 24
Inpatient Payer Specific Charge 25
Inpatient Payer Specific Charge 26
Inpatient Payer Specific Charge 27
Inpatient Payer Specific Charge 28
Inpatient Payer Specific Charge 29
Inpatient Payer Specific Charge 3
Inpatient Payer Specific Charge 30
Inpat

In [9]:
keys = !jq -r 'keys | .[]' "/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx"

In [4]:
keys

['Discounted Cash Pricing Policy',
 'File Summary',
 'Gross Charges',
 'Inpatient De-identified Maximum Negotiated Charge',
 'Inpatient De-identified Minimum Negotiated Charge',
 'Inpatient Payer Specific Charge 1',
 'Inpatient Payer Specific Charge 10',
 'Inpatient Payer Specific Charge 11',
 'Inpatient Payer Specific Charge 12',
 'Inpatient Payer Specific Charge 13',
 'Inpatient Payer Specific Charge 14',
 'Inpatient Payer Specific Charge 15',
 'Inpatient Payer Specific Charge 16',
 'Inpatient Payer Specific Charge 17',
 'Inpatient Payer Specific Charge 18',
 'Inpatient Payer Specific Charge 19',
 'Inpatient Payer Specific Charge 2',
 'Inpatient Payer Specific Charge 20',
 'Inpatient Payer Specific Charge 21',
 'Inpatient Payer Specific Charge 22',
 'Inpatient Payer Specific Charge 23',
 'Inpatient Payer Specific Charge 24',
 'Inpatient Payer Specific Charge 25',
 'Inpatient Payer Specific Charge 26',
 'Inpatient Payer Specific Charge 27',
 'Inpatient Payer Specific Charge 28',
 'Inp

In [24]:
!jq '.["Inpatient Payer Specific Charge 26"] | .[0:100]' "/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx"

[
  {
    "Payer": "SAGAMORE",
    "MS-DRG": "020",
    "Description": "Intracranial Vascular Procedures With Principal Diagnosis Hemorrhage With Mcc",
    "Payer Specific Negotiated Charge": 744765
  },
  {
    "Payer": "SAGAMORE",
    "MS-DRG": "025",
    "Description": "Craniotomy And Endovascular Intracranial Procedures With Mcc",
    "Payer Specific Negotiated Charge": 98162
  },
  {
    "Payer": "SAGAMORE",
    "MS-DRG": "145",
    "Description": "Other Ear, Nose, Mouth And Throat O.R. Procedures Without Cc/Mcc",
    "Payer Specific Negotiated Charge": 2859
  },
  {
    "Payer": "SAGAMORE",
    "MS-DRG": "177",
    "Description": "Respiratory Infections And Inflammations With Mcc",
    "Payer Specific Negotiated Charge": 22349
  },
  {
    "Payer": "SAGAMORE",
    "MS-DRG": "207",
    "Description": "Respiratory System Diagnosis With Ventilator Support >96 Hours",
    "Payer Specific Negotiated Charge": 471091
  },
  {
    "Payer": "SAGAMORE",
    "MS-DRG": "247",
    "Descriptio

In [25]:
!jq '.["Outpatient Payer Specific Charge 8"] | .[0:100]' "/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx"

[
  {
    "Payer": "AETNA",
    "APC": "1505",
    "Description": "New Technology - Level 5 ($301 - $400)",
    "Payer Specific Negotiated Charge": 2114
  },
  {
    "Payer": "AETNA",
    "APC": "1506",
    "Description": "New Technology - Level 6 ($401 - $500)",
    "Payer Specific Negotiated Charge": 832
  },
  {
    "Payer": "AETNA",
    "APC": "5012",
    "Description": "Clinic Visits and Related Services",
    "Payer Specific Negotiated Charge": 177
  },
  {
    "Payer": "AETNA",
    "APC": "5021",
    "Description": "Level 1 Type A ED Visits",
    "Payer Specific Negotiated Charge": 460
  },
  {
    "Payer": "AETNA",
    "APC": "5022",
    "Description": "Level 2 Type A ED Visits",
    "Payer Specific Negotiated Charge": 766
  },
  {
    "Payer": "AETNA",
    "APC": "5023",
    "Description": "Level 3 Type A ED Visits",
    "Payer Specific Negotiated Charge": 2088
  },
  {
    "Payer": "AETNA",
    "APC": "5024",
    "Description": "Level 4 Type A ED Visits",
    "Payer Specific 

In [36]:
import subprocess
from pathlib import Path

def convert_ashx_to_json_by_key(ashx_file, key, output_dir):
    tmp_file_path = output_dir / f'{key}.tmp'
    print('input:', ashx_file)
    print('tmp_file', tmp_file_path)
    with open(tmp_file_path, 'w') as tmp_file:
        subprocess.run(['jq', f'.["{key}"][]', str(ashx_file)], stdout=tmp_file)
    json_file = output_dir / f'{key}.json'
    with open(json_file, 'w') as json_file:
        subprocess.run(['jq', '-s', '.', str(tmp_file_path)], stdout=json_file)
    print('output:', json_file)



In [37]:
keys[5:]

['Inpatient Payer Specific Charge 1',
 'Inpatient Payer Specific Charge 10',
 'Inpatient Payer Specific Charge 11',
 'Inpatient Payer Specific Charge 12',
 'Inpatient Payer Specific Charge 13',
 'Inpatient Payer Specific Charge 14',
 'Inpatient Payer Specific Charge 15',
 'Inpatient Payer Specific Charge 16',
 'Inpatient Payer Specific Charge 17',
 'Inpatient Payer Specific Charge 18',
 'Inpatient Payer Specific Charge 19',
 'Inpatient Payer Specific Charge 2',
 'Inpatient Payer Specific Charge 20',
 'Inpatient Payer Specific Charge 21',
 'Inpatient Payer Specific Charge 22',
 'Inpatient Payer Specific Charge 23',
 'Inpatient Payer Specific Charge 24',
 'Inpatient Payer Specific Charge 25',
 'Inpatient Payer Specific Charge 26',
 'Inpatient Payer Specific Charge 27',
 'Inpatient Payer Specific Charge 28',
 'Inpatient Payer Specific Charge 29',
 'Inpatient Payer Specific Charge 3',
 'Inpatient Payer Specific Charge 30',
 'Inpatient Payer Specific Charge 31',
 'Inpatient Payer Specific C

In [39]:
output_dir = Path('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center')
output_dir.mkdir(exist_ok=True)

for key in keys[5:]:
  print(key)
  convert_ashx_to_json_by_key(Path('/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx'), key, output_dir)
  

Inpatient Payer Specific Charge 1
input: /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
tmp_file /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 1.tmp
output: <_io.TextIOWrapper name='/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 1.json' mode='w' encoding='UTF-8'>
Inpatient Payer Specific Charge 10
input: /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
tmp_file /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 10.tmp
output: <_io.TextIOWrapper name='/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 10.json' mode='w' encoding='UTF-8'>
Inpatient Payer Specific Charge 11
input: /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
tmp_file /Users/me/data/payless_health/EIN_350868085

In [34]:
keys[2:5]

['Gross Charges',
 'Inpatient De-identified Maximum Negotiated Charge',
 'Inpatient De-identified Minimum Negotiated Charge']

In [41]:
output_dir_self_pay = Path('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums')
output_dir_self_pay.mkdir(exist_ok=True)

for key in keys[:5]:
  convert_ashx_to_json_by_key(Path('/tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx'), key, output_dir_self_pay)

input: /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
tmp_file /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums/Discounted Cash Pricing Policy.tmp
output: <_io.TextIOWrapper name='/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums/Discounted Cash Pricing Policy.json' mode='w' encoding='UTF-8'>
input: /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
tmp_file /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums/File Summary.tmp
output: <_io.TextIOWrapper name='/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums/File Summary.json' mode='w' encoding='UTF-8'>
input: /tmp/350868085_parkview-regional-medical-center_standardcharges-(1).ashx
tmp_file /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-min

In [43]:
%%sql 
SELECT * FROM read_json_auto('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums/Gross Charges.json', records=true, maximum_object_size=500000000)

CDM/EAP                                        Description  \
0      01240002                    HC GENERAL ANESTH EA ADD 30 MNS   
1      01240001                     HC GENERAL ANESTH INITIAL HOUR   
2      01240010                      HC IV SEDATION EA ADD 30 MINS   
3      01240009                        HC IV SEDATION INITIAL HOUR   
4      01240008                  HC MONITORED ANESTH EA ADD 30 MNS   
...         ...                                                ...   
13912     11700                            ZOLPIDEM 10 MG ORAL TAB   
13913     36987                         ZONISAMIDE  25 MG ORAL CAP   
13914    800568          ZONISAMIDE 10 MG/ML SUSPENSION (ANY SIZE)   
13915     27780                         ZONISAMIDE 100 MG ORAL CAP   
13916     76518  ZOSTER VACCINE LIVE (PF) 19400 UNIT/0.65 ML SU...   

      Parkview Regional Medical Center  \
0                                 1056   
1                                 2110   
2                                  383   
3                                  764   
4                                  538   
...                                ...   
13912                             1.09   
13913                             1.28   
13914                            33.14   
13915                              1.4   
13916                          1067.71   

      Discount Cash Price (see discount cash policy for add'l information)  \
0                                                   1056                     
1                                                   2110                     
2                                                    383                     
3                                                    764                     
4                                                    538                     
...                                                  ...                     
13912                                               1.09                     
13913                                               1.28                     
13914                                              33.14                     
13915                                                1.4                     
13916                                            1067.71                     

      CPT/HCPCS          NDC  
0           NaN          NaN  
1           NaN          NaN  
2           NaN          NaN  
3           NaN          NaN  
4           NaN          NaN  
...         ...          ...  
13912            13668000801  
13913            68001024200  
13914            00000007875  
13915            68001024400  
13916     90736  00006496300  

[13917 rows x 6 columns]

In [44]:
%%sql 
SELECT * FROM read_json_auto('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/self-pay-minimums-maximums/Inpatient De-identified Maximum Negotiated Charge.json', records=true, maximum_object_size=500000000)

MS-DRG                                        Description  \
0      003  Ecmo Or Tracheostomy With Mv >96 Hours Or Prin...   
1      004  Tracheostomy With Mv >96 Hours Or Principal Di...   
2      011  Tracheostomy For Face, Mouth And Neck Diagnose...   
3      012  Tracheostomy For Face, Mouth And Neck Diagnose...   
4      013  Tracheostomy For Face, Mouth And Neck Diagnose...   
..     ...                                                ...   
676    983  Extensive O.R. Procedures Unrelated To Princip...   
677    987  Non-Extensive O.R. Procedures Unrelated To Pri...   
678    988  Non-Extensive O.R. Procedures Unrelated To Pri...   
679    989  Non-Extensive O.R. Procedures Unrelated To Pri...   
680    998  Principal Diagnosis Invalid As Discharge Diagn...   

     De-Identified Maximum Negotiated Charge  
0                                    1511265  
1                                     693589  
2                                     241335  
3                                     118201  
4                                      43267  
..                                       ...  
676                                    42287  
677                                   270888  
678                                    63067  
679                                    25680  
680                                    12478  

[681 rows x 3 columns]

In [15]:
%%sql 
SELECT * FROM read_json_auto('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 9.json', records=true, maximum_object_size=500000000)

Payer   APC                             Description  \
0    CIGNA  1505  New Technology - Level 5 ($301 - $400)   
1    CIGNA  1506  New Technology - Level 6 ($401 - $500)   
2    CIGNA  5012      Clinic Visits and Related Services   
3    CIGNA  5021                Level 1 Type A ED Visits   
4    CIGNA  5022                Level 2 Type A ED Visits   
..     ...   ...                                     ...   
119  CIGNA  N804                    Mammography Services   
120  CIGNA  N811                      Observation, Other   
121  CIGNA  N900                    Non-covered Services   
122  CIGNA  N902                       Packaged Services   
123  CIGNA  N905                  Not Recognized by OPPS   

     Payer Specific Negotiated Charge  
0                                 564  
1                                 726  
2                                 614  
3                                 402  
4                                 670  
..                                ...  
119                               341  
120                              1136  
121                                46  
122                              2275  
123                                62  

[124 rows x 4 columns]

In [32]:
%%sql 
SELECT "Payer" as payer,
       "APC" as billing_code,
       "Description" as description,
       "Payer Specific Negotiated Charge" as negotiated_rate 
FROM read_json_auto('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 9.json', records=true, maximum_object_size=500000000)

payer billing_code                             description  \
0    CIGNA         1505  New Technology - Level 5 ($301 - $400)   
1    CIGNA         1506  New Technology - Level 6 ($401 - $500)   
2    CIGNA         5012      Clinic Visits and Related Services   
3    CIGNA         5021                Level 1 Type A ED Visits   
4    CIGNA         5022                Level 2 Type A ED Visits   
..     ...          ...                                     ...   
119  CIGNA         N804                    Mammography Services   
120  CIGNA         N811                      Observation, Other   
121  CIGNA         N900                    Non-covered Services   
122  CIGNA         N902                       Packaged Services   
123  CIGNA         N905                  Not Recognized by OPPS   

     negotiated_rate  
0                564  
1                726  
2                614  
3                402  
4                670  
..               ...  
119              341  
120             1136  
121               46  
122             2275  
123               62  

[124 rows x 4 columns]

In [18]:
%%sql 
SELECT * FROM read_json_auto('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 1.json', records=true, maximum_object_size=500000000)

Payer MS-DRG                                        Description  \
0    AETNA    003  Ecmo Or Tracheostomy With Mv >96 Hours Or Prin...   
1    AETNA    024  Craniotomy With Major Device Implant Or Acute ...   
2    AETNA    027  Craniotomy And Endovascular Intracranial Proce...   
3    AETNA    029  Spinal Procedures With Cc Or Spinal Neurostimu...   
4    AETNA    039             Extracranial Procedures Without Cc/Mcc   
..     ...    ...                                                ...   
213  AETNA    958  Other O.R. Procedures For Multiple Significant...   
214  AETNA    964          Other Multiple Significant Trauma With Cc   
215  AETNA    974          Hiv With Major Related Condition With Mcc   
216  AETNA    982  Extensive O.R. Procedures Unrelated To Princip...   
217  AETNA    988  Non-Extensive O.R. Procedures Unrelated To Pri...   

     Payer Specific Negotiated Charge  
0                               86158  
1                              170268  
2                              189078  
3                              172548  
4                               85582  
..                                ...  
213                            145562  
214                             20503  
215                             25441  
216                             35115  
217                             63067  

[218 rows x 4 columns]

In [33]:
%%sql 
SELECT "Payer" as payer,
       "MS-DRG" as billing_code,
       "Description" as description,
       "Payer Specific Negotiated Charge" as negotiated_rate 
FROM read_json_auto('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 1.json', records=true, maximum_object_size=500000000)

payer billing_code                                        description  \
0    AETNA          003  Ecmo Or Tracheostomy With Mv >96 Hours Or Prin...   
1    AETNA          024  Craniotomy With Major Device Implant Or Acute ...   
2    AETNA          027  Craniotomy And Endovascular Intracranial Proce...   
3    AETNA          029  Spinal Procedures With Cc Or Spinal Neurostimu...   
4    AETNA          039             Extracranial Procedures Without Cc/Mcc   
..     ...          ...                                                ...   
213  AETNA          958  Other O.R. Procedures For Multiple Significant...   
214  AETNA          964          Other Multiple Significant Trauma With Cc   
215  AETNA          974          Hiv With Major Related Condition With Mcc   
216  AETNA          982  Extensive O.R. Procedures Unrelated To Princip...   
217  AETNA          988  Non-Extensive O.R. Procedures Unrelated To Pri...   

     negotiated_rate  
0              86158  
1             170268  
2             189078  
3             172548  
4              85582  
..               ...  
213           145562  
214            20503  
215            25441  
216            35115  
217            63067  

[218 rows x 4 columns]

In [4]:
%%sql 
SELECT "Payer" as payer,
       "MS-DRG" as billing_code,
       "Description" as description,
       "Payer Specific Negotiated Charge" as negotiated_rate 
FROM read_json('/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 1.json', 
               format='array',
               columns={'Payer': 'VARCHAR',
                        'MS-DRG': 'VARCHAR',
                        'Description': 'VARCHAR',
                        'Payer Specific Negotiated Charge': 'DOUBLE'})

payer billing_code                                        description  \
0    AETNA          003  Ecmo Or Tracheostomy With Mv >96 Hours Or Prin...   
1    AETNA          024  Craniotomy With Major Device Implant Or Acute ...   
2    AETNA          027  Craniotomy And Endovascular Intracranial Proce...   
3    AETNA          029  Spinal Procedures With Cc Or Spinal Neurostimu...   
4    AETNA          039             Extracranial Procedures Without Cc/Mcc   
..     ...          ...                                                ...   
213  AETNA          958  Other O.R. Procedures For Multiple Significant...   
214  AETNA          964          Other Multiple Significant Trauma With Cc   
215  AETNA          974          Hiv With Major Related Condition With Mcc   
216  AETNA          982  Extensive O.R. Procedures Unrelated To Princip...   
217  AETNA          988  Non-Extensive O.R. Procedures Unrelated To Pri...   

     negotiated_rate  
0            86158.0  
1           170268.0  
2           189078.0  
3           172548.0  
4            85582.0  
..               ...  
213         145562.0  
214          20503.0  
215          25441.0  
216          35115.0  
217          63067.0  

[218 rows x 4 columns]

In [25]:
import duckdb
from tqdm.notebook import tqdm

for file in tqdm([f for f in output_dir.glob('*.json') if 'Inpatient' in f.name]):
  file_path = file.absolute().as_posix()
  print(file_path)
  file_path_output = file.with_suffix('.parquet').absolute().as_posix()
  print(file_path_output)
  sql = f"""COPY (
            SELECT "Payer" as payer,
                  "MS-DRG" as billing_code,
                  "Description" as description,
                  "Payer Specific Negotiated Charge" as negotiated_rate  
            FROM read_json('{file_path}',
                          format='array',
                          columns={{'Payer': 'VARCHAR',
                                  'MS-DRG': 'VARCHAR',
                                  'Description': 'VARCHAR',
                                  'Payer Specific Negotiated Charge': 'DOUBLE'}})
            ) TO '{file_path_output}' (COMPRESSION ZSTD)
  """
  duckdb.sql(sql)

  0%|          | 0/60 [00:00<?, ?it/s]

/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 3.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 3.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 28.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 28.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 12.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 12.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 45.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Inpatient Payer Specific Charge 45.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-r

In [26]:

for file in tqdm([f for f in output_dir.glob('*.json') if 'Outpatient' in f.name]):
  file_path = file.absolute().as_posix()
  print(file_path)
  file_path_output = file.with_suffix('.parquet').absolute().as_posix()
  print(file_path_output)
  sql = f"""COPY (
            SELECT "Payer" as payer,
                  "MS-DRG" as billing_code,
                  "Description" as description,
                  "Payer Specific Negotiated Charge" as negotiated_rate  
            FROM read_json('{file_path}',
                          format='array',
                          columns={{'Payer': 'VARCHAR',
                                  'MS-DRG': 'VARCHAR',
                                  'Description': 'VARCHAR',
                                  'Payer Specific Negotiated Charge': 'DOUBLE'}})
            ) TO '{file_path_output}' (COMPRESSION ZSTD)
  """
  duckdb.sql(sql)

  0%|          | 0/75 [00:00<?, ?it/s]

/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 52.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 52.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 44.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 44.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 3.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 3.parquet
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 13.json
/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/Outpatient Payer Specific Charge 13.parquet
/Users/me/data/payless_health/EIN_350868085_pa

In [27]:
%%sql 
SELECT DISTINCT description FROM '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet'

description
0    Other Antepartum Diagnoses Without O.R. Proced...
1    Septicemia Or Severe Sepsis Without Mv >96 Hou...
2    Craniotomy With Major Device Implant Or Acute ...
3               Extracranial Procedures Without Cc/Mcc
4    Degenerative Nervous System Disorders Without Mcc
..                                                 ...
834               Skin Grafts For Injuries With Cc/Mcc
835                                Non-Extensive Burns
836        Hiv With Or Without Other Related Condition
837  Chemotherapy Without Acute Leukemia As Seconda...
838        Hiv With Extensive O.R. Procedures With Mcc

[839 rows x 1 columns]

In [46]:
%%sql 
SELECT DISTINCT billing_code, description, payer FROM '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet'

billing_code                                        description  \
0              027  Craniotomy And Endovascular Intracranial Proce...   
1              178   Respiratory Infections And Inflammations With Cc   
2              204                     Respiratory Signs And Symptoms   
3              236  Coronary Bypass Without Cardiac Catheterizatio...   
4              280  Acute Myocardial Infarction, Discharged Alive ...   
...            ...                                                ...   
10973          815  Reticuloendothelial And Immunity Disorders Wit...   
10974          946                      Rehabilitation Without Cc/Mcc   
10975          870      Septicemia Or Severe Sepsis With Mv >96 Hours   
10976          305                           Hypertension Without Mcc   
10977          735  Pelvic Evisceration, Radical Hysterectomy And ...   

                              payer  
0             PLAIN CHURCH-ADAMS CO  
1             PLAIN CHURCH-ADAMS CO  
2             PLAIN CHURCH-ADAMS CO  
3             PLAIN CHURCH-ADAMS CO  
4             PLAIN CHURCH-ADAMS CO  
...                             ...  
10973  MEDICARE MANAGED CARE-HUMANA  
10974  MEDICARE MANAGED CARE-HUMANA  
10975  MEDICARE MANAGED CARE-HUMANA  
10976  MEDICARE MANAGED CARE-HUMANA  
10977                           PHP  

[10978 rows x 3 columns]

In [47]:
%%sql 
SELECT * FROM '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet'

payer billing_code  \
0                             PHP PIZZA HUT          831   
1                             PHP PIZZA HUT          872   
2      MEDICAID MANAGED CARE-CARESOURCE HHW          100   
3      MEDICAID MANAGED CARE-CARESOURCE HHW          101   
4      MEDICAID MANAGED CARE-CARESOURCE HHW          176   
...                                     ...          ...   
11137         MEDICAID MANAGED CARE-UHC HHC          832   
11138         MEDICAID MANAGED CARE-UHC HHC          866   
11139         MEDICAID MANAGED CARE-UHC HHC          882   
11140         MEDICAID MANAGED CARE-UHC HHC          885   
11141         MEDICAID MANAGED CARE-UHC HHC          897   

                                             description  negotiated_rate  
0      Other Antepartum Diagnoses Without O.R. Proced...           8445.0  
1      Septicemia Or Severe Sepsis Without Mv >96 Hou...          12772.0  
2                                      Seizures With Mcc           7370.0  
3                                   Seizures Without Mcc           3699.0  
4                         Pulmonary Embolism Without Mcc           5356.0  
...                                                  ...              ...  
11137  Other Antepartum Diagnoses Without O.R. Proced...           3688.0  
11138                          Viral Illness Without Mcc           2182.0  
11139                         Neuroses Except Depressive           2980.0  
11140                                          Psychoses           3878.0  
11141  Alcohol, Drug Abuse Or Dependence Without Reha...           8591.0  

[11142 rows x 4 columns]

# Claude Prompt 

````
```
WITH max_min_charges AS (
  SELECT
    cpt_hcpcs_drg AS billing_code, 
    description,
    MIN(negotiated_rate) AS min_charge,
    MAX(negotiated_rate) AS max_charge
  FROM '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet'
  WHERE negotiated_rate > 158
  GROUP BY cpt_hcpcs_drg, description
),

min_max_names AS (
  SELECT
    mmc.billing_code,
    mmc.description, 
    mmc.min_charge,
    mmc.max_charge,
    FIRST_VALUE(insurance_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate ASC) AS name_minimum,
    FIRST_VALUE(insurance_name) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate DESC) AS name_maximum
  FROM max_min_charges mmc
  JOIN '/Users/me/data/payless_health/351955872_indiana-university-health-inc._standardcharges.parquet' sc
  ON mmc.billing_code = sc.cpt_hcpcs_drg AND mmc.description = sc.description
)

SELECT DISTINCT
  billing_code,
  description,
  min_charge,
  max_charge,
  name_minimum,
  name_maximum
FROM min_max_names
``` 

Please use the above query and rewrite it for the file `/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet` with the following column names `payer, billing_code, description, negotiated_rate`
````

In [50]:
%%sql 
WITH max_min_charges AS (
  SELECT
    billing_code, 
    description,
    MIN(negotiated_rate) AS min_charge,
    MAX(negotiated_rate) AS max_charge
  FROM '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet'
  -- WHERE negotiated_rate > 0
  -- WHERE negotiated_rate > 157
  GROUP BY billing_code, description
),

min_max_names AS (
  SELECT
    mmc.billing_code,
    mmc.description,
    mmc.min_charge, 
    mmc.max_charge,
    FIRST_VALUE(payer) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate ASC) AS name_minimum,  
    FIRST_VALUE(payer) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate DESC) AS name_maximum
  FROM max_min_charges mmc
  JOIN '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet' sc
  ON mmc.billing_code = sc.billing_code AND mmc.description = sc.description
)

SELECT DISTINCT
  billing_code,
  description,
  min_charge,
  max_charge,
  name_minimum,
  name_maximum
FROM min_max_names

billing_code                                        description  \
0            028                         Spinal Procedures With Mcc   
1            054                  Nervous System Neoplasms With Mcc   
2            088                                Concussion With Mcc   
3            103                              Headaches Without Mcc   
4            135           Sinus And Mastoid Procedures With Cc/Mcc   
..           ...                                                ...   
676          747  Vagina, Cervix And Vulva Procedures Without Cc...   
677          786    Cesarean Section Without Sterilization With Mcc   
678          223  Cardiac Defibrillator Implant With Cardiac Cat...   
679          099  Non-Bacterial Infection Of Nervous System Exce...   
680          374                      Digestive Malignancy With Mcc   

     min_charge  max_charge                          name_minimum  \
0       17451.0    289868.0                          OPTUM VA CCN   
1        4929.0     64851.0                          OPTUM VA CCN   
2       10880.0     10880.0             MEDICARE MANAGED CARE-UHC   
3        2853.0     25931.0                          OPTUM VA CCN   
4       40792.0     40792.0                                   UHC   
..          ...         ...                                   ...   
676      7192.0     95638.0          MEDICARE MANAGED CARE-HUMANA   
677     11833.0     84683.0          MEDICARE MANAGED CARE-HUMANA   
678     40623.0    163401.0             MEDICARE MANAGED CARE-UHC   
679      6530.0     29104.0  MEDICAID MANAGED CARE-CARESOURCE HIP   
680      3109.0     60356.0         MEDICAID MANAGED CARE-BUCKEYE   

                   name_maximum  
0    ANTHEM HMO/PPO/TRADITIONAL  
1    ANTHEM HMO/PPO/TRADITIONAL  
2     MEDICARE MANAGED CARE-UHC  
3                         CIGNA  
4                           UHC  
..                          ...  
676                         PHP  
677         SIGNATURE CARE-EACS  
678  ANTHEM HMO/PPO/TRADITIONAL  
679  ANTHEM HMO/PPO/TRADITIONAL  
680                         PHP  

[681 rows x 6 columns]

In [51]:
%%sql 
COPY (
WITH max_min_charges AS (
  SELECT
    billing_code, 
    description,
    MIN(negotiated_rate) AS min_charge,
    MAX(negotiated_rate) AS max_charge
  FROM '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet'
  -- WHERE negotiated_rate > 0
  -- WHERE negotiated_rate > 157
  GROUP BY billing_code, description
),

min_max_names AS (
  SELECT
    mmc.billing_code,
    mmc.description,
    mmc.min_charge, 
    mmc.max_charge,
    FIRST_VALUE(payer) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate ASC) AS name_minimum,  
    FIRST_VALUE(payer) OVER (PARTITION BY mmc.billing_code, mmc.description ORDER BY negotiated_rate DESC) AS name_maximum
  FROM max_min_charges mmc
  JOIN '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center/*.parquet' sc
  ON mmc.billing_code = sc.billing_code AND mmc.description = sc.description
)

SELECT DISTINCT
  billing_code,
  description,
  min_charge,
  max_charge,
  name_minimum,
  name_maximum
FROM min_max_names
) TO '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center_analysis.parquet' (FORMAT 'parquet');

Empty DataFrame
Columns: [Success]
Index: []

In [52]:
%%sql 
SELECT * FROM '/Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center_analysis.parquet'

billing_code                                        description  \
0            028                         Spinal Procedures With Mcc   
1            054                  Nervous System Neoplasms With Mcc   
2            088                                Concussion With Mcc   
3            103                              Headaches Without Mcc   
4            135           Sinus And Mastoid Procedures With Cc/Mcc   
..           ...                                                ...   
676          333                           Rectal Resection With Cc   
677          747  Vagina, Cervix And Vulva Procedures Without Cc...   
678          786    Cesarean Section Without Sterilization With Mcc   
679          223  Cardiac Defibrillator Implant With Cardiac Cat...   
680          374                      Digestive Malignancy With Mcc   

     min_charge  max_charge                   name_minimum  \
0       17451.0    289868.0                   OPTUM VA CCN   
1        4929.0     64851.0                   OPTUM VA CCN   
2       10880.0     10880.0      MEDICARE MANAGED CARE-UHC   
3        2853.0     25931.0                   OPTUM VA CCN   
4       40792.0     40792.0                            UHC   
..          ...         ...                            ...   
676     15570.0     89589.0    MEDICARE MANAGED CARE-AETNA   
677      7192.0     95638.0   MEDICARE MANAGED CARE-HUMANA   
678     11833.0     84683.0   MEDICARE MANAGED CARE-HUMANA   
679     40623.0    163401.0      MEDICARE MANAGED CARE-UHC   
680      3109.0     60356.0  MEDICAID MANAGED CARE-BUCKEYE   

                   name_maximum  
0    ANTHEM HMO/PPO/TRADITIONAL  
1    ANTHEM HMO/PPO/TRADITIONAL  
2     MEDICARE MANAGED CARE-UHC  
3                         CIGNA  
4                           UHC  
..                          ...  
676                         UHC  
677                         PHP  
678         SIGNATURE CARE-EACS  
679  ANTHEM HMO/PPO/TRADITIONAL  
680                         PHP  

[681 rows x 6 columns]

In [53]:
!ls -lh /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center_analysis.parquet

-rw-r--r--  1 me  staff    25K Sep  7 16:59 /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center_analysis.parquet


In [54]:
!cp /Users/me/data/payless_health/EIN_350868085_parkview-regional-medical-center_analysis.parquet /Users/me/projects/payless.health/docs/public/data/

# Claude Prompt to make visualization

````
```
meta:
  title: Hospital Charge Data
  description: An interactive dashboard of hospital charge data 

data:
  charges: {file: data/231689692_Lehigh_Valley_Hospital_StandardCharges_analysis.parquet}
  dd: 
    - { u: 0, v: 0 }
    - { u: 500000, v: 500000 }

hconcat:
- vconcat:
  - hconcat:
    - input: menu
      label: Insurance_minimum
      as: $query  
      from: charges
      column: name_minimum
    - input: search 
      label: description
      as: $query
      from: charges
      column: description
      type: contains
  - vspace: 10  
  - plot:
    - mark: dot
      data: {from: charges, filterBy: $query}
      x: min_charge
      y: max_charge
      opacity: 0.1
      fill: name_minimum
      r: 2      
    - mark: regressionY
      data: {from: charges, filterBy: $query}
      x: min_charge  
      y: max_charge
    - select: intervalXY
      as: $query
      brush: { fillOpacity: 0, stroke: black }    
    - mark: lineY
      data: { from: dd }
      x: u
      y: v
      stroke: red
    margins: { left: 60, top: 20, right: 60 }
    xyDomain: Fixed 
    width: 590
    height: 350
    yDomain: [0, 100000]
    xDomain: [0, 100000]
    yGrid: true
    xGrid: true
  - vspace: 5
  - input: table 
    from: charges
    filterBy: $query
    maxWidth: 700
    columns:
      - billing_code
      - description 
      - min_charge
      - max_charge
      - name_minimum
      - name_maximum
    width:  
      billing_code: 60
      description: 200
      min_charge: 90
      max_charge: 90
      name_maximum: 120
      name_minimum: 120
  ```

  Please use the above YAML file and rewrite it for the file `data/EIN_350868085_parkview-regional-medical-center_analysis.parquet` that has the following column names: `billing_code, description, min_charge, max_charge, name_minimum, name_maximum` 
  ````
